PEP 836: JIT Go Brrr: The Path to a Supported JIT Compiler for CPython (#5016)
python/peps: peps/pep-0836.rst

.. _jit-learnings:

Learnings
=========

JIT projects evolve over their lifespan, as seen for example in CRuby which has
seen multiple JIT compilers. CPython's JIT is no different.

CPython's JIT compiler has areas to improve. To be sustainable in the
long-term, meet our performance goals, and continue fostering community
engagement, certain tradeoffs are required. Tooling, compatibility with other
JIT projects, and free-threading must be first-class citizens.

In our experience, there are several areas which we believe have been
successful:

* **The bytecode DSL and uops.** This approach lowers maintenance burden even
  in the interpreter, as repeated units of code can be shared without
  interpretive overhead, and we can reduce error-proneness when modifying the
  interpreter through bytecode validation. These should remain in CPython even
  if the current JIT is unsuccessful and asked to be removed. The uops
  themselves also form the intermediate representation for the JIT
  automatically.

* **Generating a JIT translator automatically using our own tooling.**
  CPython's JIT can automatically generate bytecode to JIT Intermediate

>   Representation (IR) rules automatically using the bytecodes DSL. Again, this

Reviewer: ```suggestion
  Representation (IR) rules using the bytecodes DSL. Again, this
```

  means most JIT translations are correct-by-construction, reducing
  error-proneness and maintenance burden. This also means the complexity of
  the JIT is

>   self-contained: adding new features to CPython generally do not need JIT

Reviewer: ```suggestion
  self-contained: new features added to CPython generally do not need JIT
```

  support unless implementers want the JIT to optimize the feature. For

>   example, the initial lazy imports pull request[#lazy-imports-pr]_ did not

Reviewer 2: The reference/link doesn't render correctly at the moment: 

<img width="598" height="122" alt="Image" src="https://github.com/user-attachments/assets/505f1239-347d-49a1-9209-ae84b6f8ad4d" />

Reviewer 3: Woops, thanks!

  require touching any JIT files apart from adding new headers to include in C.

* **A JIT optimizer that resembles the CPython interpreter.** The current JIT

>   optimizer (middle-end), analyzes type information over CPython uops. The key

Reviewer: ```suggestion
  optimizer (middle-end) analyzes type information over CPython uops. The key
```

  maintain it for a wide variety of platforms.

* **Trace recording provides some benefits naturally.** For example,
  polymorphism, speculation, dead code elimination, value recording are well
  handled in a trace recording JIT.

* **Traces are an easy starting point.** Traces don't have control-flow within
  them, making analysis simpler.

* **A community-maintained JIT.** Despite partial funding from corporate
  sources (which we are grateful for), a sizable portion of JIT work comes from
  volunteers. Breaking the JIT into understandable chunks for contributors to
  work on is an effective way of compartmentalizing complexity and encouraging
  ownership.

Conversely, we have also learned quite a bit about what has not worked for
CPython and what could be improved upon:

* **We can continue to improve community outreach and engagement.** We are
  taking active steps to onboard new members. However, continuous engagement
  with the wider community and their requests/needs is critical for the
  project. This includes talking to system distributors for example, or
  maintainers of third-party tooling, understanding their concerns, and

>   accommodating them better

Reviewer 4: ```suggestion
  accommodating them better.
```

  feedback so far suggests that LLVM itself is not always the main obstacle as
  most can provide a recent LLVM toolchain. The harder problem is depending on
  one exact LLVM version for the lifetime of a Python release, which can force
  redistributors to carry multiple LLVM versions, rely on unsupported
  toolchains, disable the JIT or maintain bespoke stencil generation workflows.
  We must codify a solution that is workable for distributors. :pep:`774` is
  one such solution but more research needs to be done to prevent each Linux
  distribution rolling their own bespoke solution.

Rationale
=========

As noted :ref:`above <current-state>`, the JIT has achieved roughly 4-12%
faster geometric mean on the pyperformance benchmark suite for measured Tier 1
platforms (see :ref:`Appendix <appendix-jit-speedup-2wk>`), with
some limitations, challenges and areas of improvement. In this next phase of
the JIT, **we want to set an initial ambitious but attainable target of at
least 20% performance improvement over the interpreter on the free-threaded
build achieved within the next 2.5 years (in other words, by Python 3.17).**

> However, we know that performance for performance sake and at the cost of

Reviewer 4: ```suggestion
However, we know that performance for performance's sake and at the cost of
```

make the next set of larger optimizations easier to implement, reason about,
test, and maintain.

As part of our plans, we plan to optimize generators/coroutines better and
improve the efficiency of calls. These high-level optimizations motivated us
towards a method-based JIT. These optimizations require seeing more of the
user's code to be effective, and the current tracing JIT in CPython cannot
achieve this without inter-trace optimization or trace stitching, which
increases the complexity and coupling with the runtime.

Further optimizations are possible. However, they do not differ much if a trace
recording or method frontend is used:

* Lock removal on free-threading
* Detecting deferred reclamation to reduce escaping sites in the JIT on
  free-threading
* Conservative unboxing of integers, floats, and small strings

To recover the optimizations tracing gives for free, we plan to explore:

* Recording extra type profiling information from the interpreter's specializer
* Path splitting (duplicating the control-flow graph)
* Cold code elimination
* Respecializing instructions in the middle-end.

One may argue that this introduces a lot of complexity in the method JIT.

> However, optimization 1 involves no changes to the interpreter and only minimal

Reviewer: The previous list isn't numbered, what are optimizations 1, 2, etc.? I'd prefer giving them short names ("type profiling", "path splitting") and eliminating the numbers.

  Reclamation (QSBR). This unlocks even more optimization opportunities that
  are not possible with immediate reclamation.
* The JIT can remove locks and atomics in the specializing adaptive interpreter
  when it detects that objects are uniquely referenced. This is a source of
  slowdown on architectures where atomics are more expensive.

**We believe that the right framing here is not the JIT *or* free-threading,
but rather, the JIT *and* free-threading**. We understand the JIT may initially
lose some performance opportunities from free threading's semantics. However,
both the JIT and free threading have much to gain. The JIT can recover all of
free threading's single-threaded performance losses and maybe even more.

.. _distribution:

A Better JIT Distribution Story
-------------------------------

We can choose to adopt either :pep:`774`'s solution or allow a range of LLVM
versions to build the JIT. This is up for more discussion and
experimentation. At minimum, feedback from popular Linux distributors must be
collected, deliberated on, and incorporated into a holistic solution. Our
current understanding of the situation is that supporting multiple LLVM

> versions is required. This should not be much additional complexity. However,

Reviewer 4: ```suggestion
versions is required. This should not add much additional complexity. However,
```

interpreter approachable or decoupled from the JIT, to documenting the workflow
for regenerating generated code and contributing changes, and to keeping the
internals documentation current. The simplification of the optimizer's
operations shipped in 3.15 (see :ref:`Learnings <jit-learnings>`)
is an example of this maintenance investment in practice. Obligations on
redistributors who build and ship the JIT are described in the
:ref:`A Better JIT Distribution Story <distribution>` section.

Backwards Compatibility
=======================

Since the JIT is an optimization and not a change to the language, its central
compatibility guarantee is that a JIT-enabled build must produce behavior
indistinguishable from a non-JIT build, just faster: same results, same
exceptions and tracebacks, and same supported introspectable state.

As :ref:`covered above <compatibility>`, we will conduct a compatibility
analysis on the top PyPI packages' test suite as a requirement to regard the
JIT as supported in CPython.

Security Implications
=====================

As stated in :pep:`744`, CPython's JIT, like all JITs, produces large amounts

> of executable data at runtime. This is an attack vector of all JIT compilers, a

Reviewer 4: ```suggestion
of executable data at runtime. This is an attack vector of all JIT compilers: a
```

[#issue-149697]_. We believe supporting this option in the generated stencils
is required for meeting our goals for security.

Finally, since :pep:`744`'s inception, multiple fuzzing projects have been
initiated to fuzz the JIT. For example, Lafleur [#la-fleur]_ has found
numerous JIT bugs that lead to crashes or wrong optimizations (mostly in the
JIT middle-end, not the backend). We will continue using these projects to fuzz
the JIT.

How to Teach This
=================

For the vast majority of Python users, the most important thing to teach about
the JIT is that there is nothing they need to do and nothing they need to watch
out for. No code should need to be rewritten to benefit from the JIT, and none
should need to be changed to remain correct.

For users who want a mental model, a short and accurate one is enough: the JIT
is an optimization layer that sits above the interpreter and compiles
frequently-executed code to machine code on the fly. It changes how fast a
program runs, not what it does. This framing is sufficient for most educational
contexts and does not require teaching the internals (for example, uops,
optimizer, code generation).

> Two audiences that need more specific guidance - redistributors and packagers.

Reviewer: ```suggestion
Two audiences need more specific guidance: redistributors and packagers.
```

Once the JIT is regarded as supported, it should not be broken catastrophically
by any new changes. However, we expect that in almost all cases, introducing a
new feature to Python will not be obstructed by a JIT, unless the contributor
explicitly wants the JIT to support their feature or optimize for it. Once
again, see for example the lazy imports initial implementation which modified
bytecode, but did not need to modify the JIT other than ``#include`` the new
headers introduced [#lazy-imports-pr]_.

Reference Implementation
========================

The current implementation for the JIT can be found in CPython's main branch,
largely in:

* ``Tools/jit/README.md``: Instructions for how to build the JIT.
* ``Python/jit.c``: The entire backend portion of the JIT compiler.
* ``Python/optimizer.c``: Part of the frontend of the JIT compiler (partially
  shared from ``Python/ceval.c``).
* ``Python/optimizer_analysis.c``: The middle-end of the JIT compiler.
* ``Python/optimizer_bytecodes.c``: The middle-end of the JIT compiler's
  optimization rules.
* ``jit_stencils.h``: An example of the JIT's build-time generated templates

>   (not currently checked into CPython repository).

Reviewer 4: ```suggestion
  (not currently checked into the CPython repository).
```

the JIT in the main branch of CPython also allows us to have a better pulse on
the needs of distributors, and means that it's easier for end users to try out
the JIT and let us know what behavior they observe and what issues they find.

Pluggable JIT Infrastructure
----------------------------

Another recurring idea is for CPython to expose a stable, general-purpose
interface for plugging in arbitrary third-party JIT compilers, rather than
maintaining one in tree. This PEP rejects this idea for the same reasons as
maintaining the JIT outside of CPython main. Introducing a pluggable JIT risks
diverting contributor effort and increases maintenance overhead. For example,
an earlier version of the JIT in 3.13 had a semi-public experimental API.
However, it leaked internal details to "users" (there were none) and made
internal JIT development more difficult. Thus, we removed it. Furthermore,
most language runtime JITs are deeply integrated with their respective
runtimes to the extent that a pluggable JIT infrastructure may not be feasible.

We agree however, that efforts that maintain a JIT outside of CPython using

> :pep:`523`, such as CinderX and TorchDynamo are commendable. We believe the

Reviewer: ```suggestion
:pep:`523`, such as CinderX and TorchDynamo, are commendable. We believe the
```

> discussion to be had for improving the pre-existing interfaces are best left to

Reviewer: ```suggestion
discussions to be had for improving the pre-existing interfaces are best left to
```

a separate PEP, and consider them out of scope for this one.

Dropping the Build-Time LLVM Requirement
----------------------------------------

This PEP does not propose changing the JIT's reliance on the LLVM toolchain at
build-time. We treat reducing build-time friction as important, but not as a
precondition for agreeing on the path outlined here. :pep:`774` proposes a
solution for removing the LLVM prerequisite but at time of submission, the
sitting Steering Council decided to defer making a decision on it until the JIT

> had paid more substantial performance gains. We would also like to keep

Reviewer 4: ```suggestion
had achieved more substantial performance gains. We would also like to keep
```

+-------------------------------------+--------------+-------------+--------------+------+-------------+
| Machine                             | Config       | Avg speedup | Result       | Days | Range       |
+=====================================+==============+=============+==============+======+=============+
| jones (M3 Pro, macOS)               | JIT+TAILCALL | 1.126x      | 12.6% faster | 9    | 1.050-1.180 |
+-------------------------------------+--------------+-------------+--------------+------+-------------+
| sulaco (AmpereOne, Linux aarch64)   | JIT          | 1.073x      | 7.3% faster  | 7    | 1.060-1.080 |
+-------------------------------------+--------------+-------------+--------------+------+-------------+
| ripley (i5-8400, Linux x86_64)      | JIT          | 1.069x      | 6.9% faster  | 9    | 1.060-1.070 |
+-------------------------------------+--------------+-------------+--------------+------+-------------+
| prometheus (Ryzen 5 3600X, Windows) | JIT+TAILCALL | 1.047x      | 4.7% faster  | 9    | 1.040-1.050 |
+-------------------------------------+--------------+-------------+--------------+------+-------------+

.. note::

>    Note that JIT+TAILCALL is used on Windows and Mac run as regular CPython

Reviewer 4: ```suggestion
   Note that JIT+TAILCALL is used on Windows and macOS runs, as regular CPython
```
